spatialdata/_core/_spatialdata.py

```diff
@@ -430,40 +450,59 @@ def add_image(
             See https://zarr.readthedocs.io/en/stable/api/storage.html for more details.
         overwrite
             If True, overwrite the element if it already exists.
-        _add_in_memory
+        _called_from_write
             Internal flag, to differentiate between an element added by the user and an element saved to disk by
             write method.
 
         Notes
         -----
         If the SpatialData object is backed by a Zarr storage, the image will be written to the Zarr storage.
         """
-        if _add_in_memory:
-            self._add_image_in_memory(name=name, image=image, overwrite=overwrite)
         if self.is_backed():
-            elem_group = self._init_add_element(name=name, element_type="images", overwrite=overwrite)
-            write_image(
-                image=self.images[name],
-                group=elem_group,
-                name=name,
-                storage_options=storage_options,
-            )
-            # reload the image from the Zarr storage so that now the element is lazy loaded, and most importantly,
-            # from the correct storage
+            des_path = os.path.join(str(self.path), "images", name)
+            if _called_from_write:
+                elem_group = self._init_add_element(name=name, element_type="images", overwrite=overwrite)
+                write_image(
+                    image=self.images[name],
+                    group=elem_group,
+                    name=name,
+                    storage_options=storage_options,
+                )
+            else:
+                self._add_image_in_memory(name=name, image=image, overwrite=overwrite)
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    store = parse_url(os.path.join(tmpdir, "data.zarr"), mode="w").store
+                    root = zarr.group(store=store)
+                    write_image(
+                        image=self.images[name],
+                        group=root,
+                        name=name,
+                        storage_options=storage_options,
+                    )
+                    src_path = os.path.join(store.path, name)
+                    assert isinstance(self.path, str)
+
+                    if os.path.isdir(des_path) and overwrite:
+                        shutil.rmtree(des_path)
+                    pathlib.Path(des_path).parent.mkdir(parents=True, exist_ok=True)
+                    shutil.move(src_path, des_path)
```

Comment:
this is the way the temporary file is created and handled by spatialdata, @ivirshup what do you think of this? I personally don't feel comfortable adding such behaviour and in general using `shutil` anywhere in the repo. I would rather constrained the behaviour and let user be explicitly about what's in memory and what on disk.

Reply:
One clarification, this code is not about choosing what is in memory and what on disk, but to deal with the following case:
1. the user load a `SpatialData` object from disk. The object contain the image `image0`, this image is a backed, lazy-loaded `xarray.DataArray`.
2. if the user modifies this object, say it doesn't a `log(1 + x)` transform, and wants to save the object into a new slot, like `image1`, this gives no problems
3. if the user instead wants to replace the `image0` slot ([like done from Wouter here](https://github.com/scverse/spatialdata/issues/137)), this requires to first write the data somewhere else, and then replacing the original location. 

Maybe there is a library for doing this kind of "atomic swap" that we can use, so we don't have to call `shutils.rmtree` implicitly. But also, if we define `des_path` right before `rmtree` (now it is defined 10 lines above), and if we do a check like `assert str(des_path.parent).endswith('.zarr')`, then I consider the call to `shutils.rmtree` to be safe. What do you think @ivirshup. Also to the attention of @kevinyamauchi

Reply:
I removed `shutils`, see comment in the discussion.